mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h

```diff
@@ -9911,48 +11015,133 @@ struct XdlopsGemmV2RewritePattern
-            ValueRange{loopKiv});
-        argB = loopKb.create<vector::TransferReadOp>(
-            loc, argType.template cast<VectorType>(), op.bufferB(),
-            ValueRange{loopKiv});
+      // Rewrite as:
+      //
+      // for(index_t k_i = 0; k_i < K; ++k_i) {
+      //   bufferAElement = a[k_i];
+      //   bufferBElement = b[k_i];
+      //   for(index_t ki_i = 0; ki_i < KRepeats; ++ki_i)
+      //     argA = &bufferAElement[ki_i * mfma_type.k_base];
+      //     argB = &bufferAElement[ki_i * mfma_type.k_base];
+      //     p_c_thread = mfma_type.template run<MPerXlops * MRepeats,
+      //                                         NPerXdlops * NRepeats,
+      //                                         AStride,
+      //                                         BStride>(argA, argB,
+      //       p_c_thread);
+      // }
+
+      // FIXME: This modified KForOuterLoop shall be removed once we can be
+      // certain KPACK logic can be applied everywhere for f16 data type.
+      int64_t KForOuterLoop;
+      if (KPack > 1) {
+        KForOuterLoop = K;
+      } else {
+        KForOuterLoop = K / k_base;
+      }
+      auto outerLoop =
+          b.create<AffineForOp>(loc, 0, KForOuterLoop, 1, op.vectorCs());
+      auto outerLoopb = OpBuilder::atBlockBegin(outerLoop.getBody());
+      auto outerLoopiv = outerLoop.getInductionVar();
+
+      Type bufferAElementType =
+          op.bufferA().getType().template cast<MemRefType>().getElementType();
+      Type bufferBElementType =
+          op.bufferB().getType().template cast<MemRefType>().getElementType();
+      Value bufferAElement = outerLoopb.create<memref::LoadOp>(
+          loc, bufferAElementType, op.bufferA(), ValueRange{outerLoopiv});
+      Value bufferBElement = outerLoopb.create<memref::LoadOp>(
+          loc, bufferBElementType, op.bufferB(), ValueRange{outerLoopiv});
+
+      auto innerLoop = outerLoopb.create<AffineForOp>(
+          loc, 0, KRepeats, 1, outerLoop.getRegionIterArgs());
+      auto innerLoopb = OpBuilder::atBlockBegin(innerLoop.getBody());
+      auto innerLoopiv = innerLoop.getInductionVar();
+
+      Value argA;
+      Value argB;
+      int64_t argTypeVectorLength =
+          (argType.isa<VectorType>())
+              ? argType.template cast<VectorType>().getShape()[0]
+              : 1;
+      if (argTypeVectorLength > 1) {
+        Value zeroOp = createZeroConstantFloatOp(innerLoopb, loc, dataType);
+
+        Value offset =
+            innerLoopb.create<MulIOp>(loc, innerLoopiv, KBaseConstantOp);
+        if (bufferAElementType.isa<VectorType>()) {
+          // bufferA/BElement loaded on LDS are vectors.
+          // argA/B to be supplied to MFMA XDLOPS are also vectors.
+          assert(bufferAElementType.isa<VectorType>());
+          assert(bufferBElementType.isa<VectorType>());
+
+          argA = innerLoopb.create<SplatOp>(loc, zeroOp, argType);
+          argB = innerLoopb.create<SplatOp>(loc, zeroOp, argType);
```

Comment:
We have zero constant creation functions in #425 (and there's a small fix to them in #441 I might want to backport related to bfloats) - FYI for your merge/rebase